### PR TITLE
USB: cdc_acm Refactor using ring buffers

### PR DESCRIPTION
**Pull request by finikorg**
Use ring buffer for TX, use usb_transfer API.

### CODE REVIEW

**Comment by jfischer-phytec-iot** [comment]
@finikorg Why is it DNM?

File: samples/subsys/usb/cdc_acm_echo/src/main.c [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/*
```
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/cdc_acm_echo/src/main.c]: :rocket: An other CDC ACM sampel :smile: Is it possible to integrate it to `subsys/usb/cdc_acm/src/main.c` and select|configure over `subsys/usb/cdc_acm/Kconfig`?

Comment (finikorg) [reply samples/subsys/usb/cdc_acm_echo/src/main.c]: What if I remove `cdc_acm` and leave only this sample. If I use config and `ifdefs` the code would be unreadable.

Comment (jfischer-phytec-iot) [reply samples/subsys/usb/cdc_acm_echo/src/main.c]: Why not rework the code of cdc_acm sample?

File: samples/subsys/usb/cdc_acm_echo/src/main.c [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#include <string.h>
+#include <device.h>
+#include <uart.h>
+#include <zephyr.h>
+#include <ring_buffer.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(cdc_acm_echo, LOG_LEVEL_DBG);
+
+#define RING_BUF_SIZE 1024
```
Comment (jfischer-phytec-iot) [inline samples/subsys/usb/cdc_acm_echo/src/main.c]: I suggest to make it default smaller and configurable over samples/subsys/usb/cdc_acm/Kconfig.

**Comment by carlescufi** [comment]
@masz-nordic can you test this on a Nordic board, including throughput?

File: samples/subsys/usb/cdc_acm_composite/sample.yaml [hunk 3]
```diff
@@ -3,6 +3,7 @@ sample:
 tests:
   sample.usb.cdc-acm-composite:
     depends_on: usb_device gpio
+    platform_exclude: frdm_kl25z
```
Comment (carlescufi) [inline samples/subsys/usb/cdc_acm_composite/sample.yaml]: @finikorg can you try to rebuild this or add a `.conf` file?

**Comment by masz-nordic** [comment]
@carlescufi Works fine on nrf52840_pca10056. Throughput is 292 kB/s.

File: samples/subsys/usb/cdc_acm_composite/sample.yaml (hunk 3 above)
Comment (finikorg) [reply samples/subsys/usb/cdc_acm_composite/sample.yaml]: I added
```
--- a/samples/subsys/usb/cdc_acm_composite/prj.conf
+++ b/samples/subsys/usb/cdc_acm_composite/prj.conf
@@ -11,6 +11,7 @@ CONFIG_USB_DEVICE_PRODUCT="Zephyr CDC ACM sample"
 CONFIG_USB_COMPOSITE_DEVICE=y
 CONFIG_USB_CDC_ACM=y
 CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
+CONFIG_USB_CDC_ACM_RINGBUF_SIZE=512
 
```

**Comment by carlescufi** [comment]
> @carlescufi Works fine on nrf52840_pca10056. Throughput is 292 kB/s.

Thank you!

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@jfischer-phytec-iot any objections to merging this?

**Review by jfischer-no — APPROVED** [review]

**Comment by jfischer-phytec-iot** [comment]
@carlescufi I would like to test on sam and kinetis platforms, but LGTM.

**Pull request merged**